File: solutions/python/protein-translation/1/protein_translation.py
def proteins(strand):
    
    proteins = []

    start = 0
    end = 3
    
    while end <= len(strand):
        codon = strand[start:end]
        protein = translate(codon)

        if protein == "STOP":
            break

        proteins.append(protein)
        start = end
        end += 3;

    return proteins

def translate(codon):

    protein = ""

    match codon:
        case "AUG":
            protein = "Methionine"
        case "UUU":
            protein = "Phenylalanine"
        case "UUC":
            protein = "Phenylalanine"
        case "UUA":
            protein = "Leucine"
        case "UUG":
            protein = "Leucine"
        case "UCU":
            protein = "Serine"
        case "UCC":
            protein = "Serine"
        case "UCA":
            protein = "Serine"
        case "UCG":
            protein = "Serine"
        case "UAU":
            protein = "Tyrosine"
        case "UAC":
            protein = "Tyrosine"
        case "UGU":
            protein = "Cysteine"
        case "UGC":
            protein = "Cysteine"
        case "UGG":
            protein = "Tryptophan"
        case "UAA":
            protein = "STOP"
        case "UAG":
            protein = "STOP"
        case "UGA":
            protein = "STOP"
        case _:
            protein = ""
    
    return protein

File: solutions/ruby/protein-translation/1/protein_translation.rb
module Translation
  def self.of_rna(strand)
    proteins = []

    position = 0
    codon_length = 3

    while position <= strand.length()

      codon = strand[position,codon_length]
      protein = self.translate(codon)

      if protein != ""
        if protein == "STOP"
          break
        else
          proteins.push(protein)
        end
      end
      position += codon_length
    end
    return proteins
  end

  def self.translate(codon)

    protein = ""
    case codon
    when ""
      protein = ""
    when "AUG"
      protein = "Methionine"
    when "UUU", "UUC"
      protein = "Phenylalanine"
    when "UUA", "UUG"
      protein = "Leucine"
    when "UCU", "UCC", "UCA", "UCG"
      protein = "Serine"
    when "UAU", "UAC"
      protein = "Tyrosine"
    when "UGU", "UGC"
      protein = "Cysteine"
    when "UGG"
      protein = "Tryptophan"
    when "UAA", "UAG", "UGA"
      protein = "STOP"
    else
      raise InvalidCodonError.new "InvalidCodonError"
    end

    return protein
  end
end

class InvalidCodonError < StandardError
end
